src/Build/Evaluation/Expander.cs

```diff
@@ -1470,6 +1470,10 @@ private static object LookupProperty(IPropertyProvider<T> properties, string pro
             private static object LookupProperty(IPropertyProvider<T> properties, string propertyName, int startIndex, int endIndex, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties)
             {
                 T property = properties.GetProperty(propertyName, startIndex, endIndex);
+                if (!string.IsNullOrEmpty(property?.EvaluatedValue))
```

Comment:
maybe use pattern-matching here

Reply:
Hmm, I am confused, this seems to just make a list of all properties ever read? We only need the properties that were resolved to an environment variable. I would imagine that we track it somewhere close to where we take a snapshot of the environment at the start of the build?

Reply:
Also I think we should move this logic down to line 1519, in the else block (when we have an escaped value).

Reply:
Basically here we should check whether the property we got came from the environment (see https://github.com/dotnet/msbuild/blob/058a0262c6b37fba7bb55ec03315232d5d83be6e/src/Build/Instance/ProjectInstance.cs#L1482)

Reply:
and yes, move this new if block into the else block, where we've verified that the property is not null. here is too early.

Reply:
https://github.com/dotnet/msbuild/blob/058a0262c6b37fba7bb55ec03315232d5d83be6e/src/Build/Evaluation/Evaluator.cs#L1200

Reply:
by pattern-matching I mean:

```
if (property?.EvaluatedValue is { Length: > 0 } evaluatedValue) { ... }
```

Reply:
What's the added value of pattern matching here of string.IsNullOrEmpty?

Reply:
worried that if the `EvaluatedValue` getter does anything expensive, or is not idempotent, we'll be calling it twice. It might be benign or not. I prefer to only call properties once and reuse, just to be safe in general.